In [9]:
import pandas as pd 
import numpy as np 


In [11]:
df=pd.read_csv("/Users/shreeshnadgouda/Downloads/resolved_queries.csv")
df1=pd.read_csv("/Users/shreeshnadgouda/Downloads/new_queries.csv")

In [12]:
df.head(10)


Query_ID                    Pre_Resolved_Query
0         1     Unable to connect to the internet
1         2        Payment failed during checkout
2         3     App crashes when opening settings
3         4   Forgot password and unable to reset
4         5  Unable to upload files to the server

In [13]:
df1.head(10)

Variation_Query  Matches_With_Query_ID
0           Unabel to conect to the internet                      1
1                  Can’t connect to internet                      1
2                        Intenet not working                      1
3               Payment failed while chekout                      2
4  Payment did not go through during chckout                      2
5                 Payment issue at check out                      2
6   Application crashes when opening setings                      3
7           App crash when going to settings                      3
8           Settings cause the app to chrash                      3
9              Forgot passwrd and cant reset                      4

In [2]:
!pip install fuzzywuzzy
!pip install python-Levenshtein


  Obtaining dependency information for fuzzywuzzy from https://files.pythonhosted.org/packages/43/ff/74f23998ad2f93b945c0309f825be92e04e0348e062026998b5eefef4c33/fuzzywuzzy-0.18.0-py2.py3-none-any.whl.metadata
  Obtaining dependency information for python-Levenshtein from https://files.pythonhosted.org/packages/72/8e/559c539e76bc0b1defec3da39a047fe151258efc9b215bf41db41e2c7922/python_Levenshtein-0.25.1-py3-none-any.whl.metadata
  Obtaining dependency information for Levenshtein==0.25.1 from https://files.pythonhosted.org/packages/58/4d/ebb238a39da4877493a41fc41763de3eb006fad13a531947f64f33622cfe/Levenshtein-0.25.1-cp310-cp310-macosx_11_0_arm64.whl.metadata
  Obtaining dependency information for rapidfuzz<4.0.0,>=3.8.0 from https://files.pythonhosted.org/packages/3d/bc/38bd009fef815f2f2fa726019f0cc16ae600b7521acccb18a7afae63dbb6/rapidfuzz-3.9.7-cp310-cp310-macosx_11_0_arm64.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.7/103.7 kB 1.7 MB/s eta 0:00:0000:01
   ━━━━━━━━━━━━━

In [4]:
!pip install sentence-transformers


  Obtaining dependency information for sentence-transformers from https://files.pythonhosted.org/packages/90/b4/52b8205f24172f2429cacf04bac324414f16b61d64e79c787c9ce2385586/sentence_transformers-3.1.0-py3-none-any.whl.metadata
  Obtaining dependency information for transformers<5.0.0,>=4.38.0 from https://files.pythonhosted.org/packages/75/35/07c9879163b603f0e464b0f6e6e628a2340cfc7cdc5ca8e7d52d776710d4/transformers-4.44.2-py3-none-any.whl.metadata
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 43.7/43.7 kB 1.4 MB/s eta 0:00:00
  Obtaining dependency information for huggingface-hub>=0.19.3 from https://files.pythonhosted.org/packages/57/28/a0b0dd3cca63908045edc300360d6cd8758d4d86eee3fd2b08f00c5a41c4/huggingface_hub-0.24.7-py3-none-any.whl.metadata
  Obtaining dependency information for safetensors>=0.4.1 from https://files.pythonhosted.org/packages/2b/9e/9648d8dbb485c40a4a0212b7537626ae440b48156cc74601ca0b7a7615e0/safetensors-0.4.5-cp310-cp310-macosx_11_0_arm64.whl.metadata
  Obtaining d

In [6]:
!pip install tf-keras


  Obtaining dependency information for tf-keras from https://files.pythonhosted.org/packages/21/8b/75f7572ec0273ed8da50bc19defe08aaaafcc15fda3407db53f49acec814/tf_keras-2.17.0-py3-none-any.whl.metadata
  Obtaining dependency information for tensorflow<2.18,>=2.17 from https://files.pythonhosted.org/packages/1f/a1/7d2042050159619a190db874913a2bc70645f8ac677d442f9aab4d29153e/tensorflow-2.17.0-cp310-cp310-macosx_12_0_arm64.whl.metadata
  Obtaining dependency information for tensorboard<2.18,>=2.17 from https://files.pythonhosted.org/packages/d4/41/dccba8c5f955bc35b6110ff78574e4e5c8226ad62f08e732096c3861309b/tensorboard-2.17.1-py3-none-any.whl.metadata
  Obtaining dependency information for keras>=3.2.0 from https://files.pythonhosted.org/packages/a3/a4/101f0f3c0b057ce150af0e8493ab7fc10b98b066b7bd81ab01e96038a268/keras-3.5.0-py3-none-any.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 9.5 MB/s eta 0:00:00ta 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 236.1/236.1

In [7]:
!pip install ipywidgets --upgrade


  Obtaining dependency information for ipywidgets from https://files.pythonhosted.org/packages/22/2d/9c0b76f2f9cc0ebede1b9371b6f317243028ed60b90705863d493bae622e/ipywidgets-8.1.5-py3-none-any.whl.metadata
  Obtaining dependency information for widgetsnbextension~=4.0.12 from https://files.pythonhosted.org/packages/21/02/88b65cc394961a60c43c70517066b6b679738caf78506a5da7b88ffcb643/widgetsnbextension-4.0.13-py3-none-any.whl.metadata
  Obtaining dependency information for jupyterlab-widgets~=3.0.12 from https://files.pythonhosted.org/packages/a9/93/858e87edc634d628e5d752ba944c2833133a28fa87bb093e6832ced36a3e/jupyterlab_widgets-3.0.13-py3-none-any.whl.metadata
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 139.8/139.8 kB 1.6 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 214.4/214.4 kB 3.4 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 17.8 MB/s eta 0:00:0000:0100:01


In [14]:
# Standard imports
import pandas as pd
import numpy as np
import re
from fuzzywuzzy import fuzz
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
from sentence_transformers import SentenceTransformer

# Assuming df is the resolved_queries and df1 is the new_queries
resolved_queries = df
new_queries = df1

# Function to preprocess text
def preprocess_text(text):
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)
    return text

# Apply preprocessing
resolved_queries['Processed_Query'] = resolved_queries['Pre_Resolved_Query'].apply(preprocess_text)
new_queries['Processed_Variation_Query'] = new_queries['Variation_Query'].apply(preprocess_text)

# Optimized fuzzy matching function (using NumPy for vectorized operations)
def fuzzy_match_batch(queries, resolved_queries, method='ratio'):
    query_list = resolved_queries['Processed_Query'].values
    match_ids = []
    match_scores = []
    
    for query in queries:
        if method == 'ratio':
            scores = np.array([fuzz.ratio(query, resolved) for resolved in query_list])
        elif method == 'partial_ratio':
            scores = np.array([fuzz.partial_ratio(query, resolved) for resolved in query_list])
        elif method == 'token_sort_ratio':
            scores = np.array([fuzz.token_sort_ratio(query, resolved) for resolved in query_list])
        elif method == 'token_set_ratio':
            scores = np.array([fuzz.token_set_ratio(query, resolved) for resolved in query_list])
        
        best_match_idx = scores.argmax()
        match_ids.append(resolved_queries.iloc[best_match_idx]['Query_ID'])
        match_scores.append(scores[best_match_idx])
    
    return match_ids, match_scores

# Apply fuzzy matching using different methods
methods = ['ratio', 'partial_ratio', 'token_sort_ratio', 'token_set_ratio']
fuzzy_results = {}

for method in methods:
    match_ids, match_scores = fuzzy_match_batch(new_queries['Processed_Variation_Query'], resolved_queries, method)
    fuzzy_results[f'Matched_Query_ID_{method.capitalize()}'] = match_ids
    fuzzy_results[f'Match_Score_{method.capitalize()}'] = match_scores

# Fuzzy matching results
fuzzy_df = pd.DataFrame(fuzzy_results)
fuzzy_df.insert(0, 'Variation_Query', new_queries['Variation_Query'])

# TF-IDF vectorization
vectorizer = TfidfVectorizer()
tfidf_resolved = vectorizer.fit_transform(resolved_queries['Processed_Query'])
tfidf_new = vectorizer.transform(new_queries['Processed_Variation_Query'])

# Cosine similarity with TF-IDF
cosine_similarities = cosine_similarity(tfidf_new, tfidf_resolved)
best_match_indices = cosine_similarities.argmax(axis=1)
best_match_scores = cosine_similarities.max(axis=1)
matched_query_ids = resolved_queries.iloc[best_match_indices]['Query_ID'].values

# TF-IDF results
tfidf_results = pd.DataFrame({
    'Variation_Query': new_queries['Variation_Query'],
    'Matched_Query_ID_TFIDF': matched_query_ids,
    'Match_Score_TFIDF': best_match_scores
})

# Use pre-trained model for semantic similarity (Sentence-BERT)
# Load Sentence-BERT model for better sentence embeddings
sbert_model = SentenceTransformer('paraphrase-MiniLM-L6-v2')

# Encode both resolved and new queries
sbert_resolved_embeddings = sbert_model.encode(resolved_queries['Processed_Query'].values)
sbert_new_embeddings = sbert_model.encode(new_queries['Processed_Variation_Query'].values)

# Compute cosine similarity for Sentence-BERT embeddings
sbert_similarities = cosine_similarity(sbert_new_embeddings, sbert_resolved_embeddings)
best_sbert_match_indices = sbert_similarities.argmax(axis=1)
best_sbert_match_scores = sbert_similarities.max(axis=1)
matched_sbert_query_ids = resolved_queries.iloc[best_sbert_match_indices]['Query_ID'].values

# Sentence-BERT results
sbert_results = pd.DataFrame({
    'Variation_Query': new_queries['Variation_Query'],
    'Matched_Query_ID_SBERT': matched_sbert_query_ids,
    'Match_Score_SBERT': best_sbert_match_scores
})

# Combine all results for comparison
final_results = pd.concat([fuzzy_df, tfidf_results[['Matched_Query_ID_TFIDF', 'Match_Score_TFIDF']], 
                           sbert_results[['Matched_Query_ID_SBERT', 'Match_Score_SBERT']]], axis=1)

# Display the final results
final_results.head(10)


/Users/shreeshnadgouda/miniforge3/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


Variation_Query  Matched_Query_ID_Ratio  \
0           Unabel to conect to the internet                       1   
1                  Can’t connect to internet                       1   
2                        Intenet not working                       1   
3               Payment failed while chekout                       2   
4  Payment did not go through during chckout                       2   
5                 Payment issue at check out                       2   
6   Application crashes when opening setings                       3   
7           App crash when going to settings                       3   
8           Settings cause the app to chrash                       3   
9              Forgot passwrd and cant reset                       4   

   Match_Score_Ratio  Matched_Query_ID_Partial_ratio  \
0                 95                               1   
1                 77                               1   
2                 42                               1   
3                 83                               2   
4                 70                               2   
5                 71                               2   
6                 88                               3   
7                 86                               3   
8                 40                               3   
9                 84                               4   

   Match_Score_Partial_ratio  Matched_Query_ID_Token_sort_ratio  \
0                         94                                  1   
1                         83                                  1   
2                         53                                  1   
3                         79                                  2   
4                         61                                  2   
5                         62                                  2   
6                         89                                  3   
7                         84                                  3   
8                         41                                  3   
9                         76                                  4   

   Match_Score_Token_sort_ratio  Matched_Query_ID_Token_set_ratio  \
0                            95                                 1   
1                            67                                 1   
2                            46                                 1   
3                            76                                 2   
4                            65                                 2   
5                            57                                 2   
6                            88                                 3   
7                            86                                 3   
8                            65                                 3   
9                            75                                 4   

   Match_Score_Token_set_ratio  Matched_Query_ID_TFIDF  Match_Score_TFIDF  \
0                           95                       1           0.839042   
1                           88                       1           0.836936   
2                           49                       1           0.000000   
3                           83                       2           0.707107   
4                           70                       2           0.707107   
5                           57                       2           0.500000   
6                           88                       3           0.774597   
7                           86                       3           0.722471   
8                           65                       3           0.508047   
9                           75                       4           0.782698   

   Matched_Query_ID_SBERT  Match_Score_SBERT  
0                       1           0.303838  
1                       1           0.862849  
2                       1           0.145931  
3                       2           0.629577  
4                       2           0.

### Output Inference:

The output contains results from several text-matching methods (Fuzzy Matching, TF-IDF, and Sentence-BERT) applied to a set of `Variation_Query` strings compared to a set of `Resolved_Queries`. Each method attempts to find the closest match between the `Variation_Query` and the `Resolved_Queries`, along with a similarity score for each match. Here is the detailed inference of the output:

#### **Row 0:**
- **Variation Query**: "Unabel to conect to the internet" (misspelled)
- **Best Match**: `Query_ID 1`
- **Scores**:
  - Fuzzy Matching (Ratio, Partial, Token Sort, Token Set): High scores (~95) indicate a strong match despite the typo.
  - TF-IDF: 0.839, which shows a good vectorized match.
  - SBERT: 0.303, much lower because SBERT is sensitive to spelling errors.
  
**Conclusion**: Fuzzy Matching performs well on this misspelled query, but SBERT struggles due to the misspelling.

#### **Row 1:**
- **Variation Query**: "Can’t connect to internet"
- **Best Match**: `Query_ID 1`
- **Scores**:
  - Fuzzy Matching: Scores range from 67 to 88, showing decent performance.
  - TF-IDF: 0.837, good match with vectorization.
  - SBERT: 0.863, a high score showing SBERT understood the semantic similarity well.

**Conclusion**: All methods, including SBERT, perform well here because the query is clear and well-formed.

#### **Row 2:**
- **Variation Query**: "Intenet not working" (misspelled "internet")
- **Best Match**: `Query_ID 1`
- **Scores**:
  - Fuzzy Matching: Lower scores (~42–53) due to the typo.
  - TF-IDF: 0, showing it did not find a good match.
  - SBERT: 0.146, indicating a poor semantic match.

**Conclusion**: Both TF-IDF and SBERT perform poorly due to the misspelling, while Fuzzy Matching handles it better.

#### **Row 3:**
- **Variation Query**: "Payment failed while chekout" (misspelled "checkout")
- **Best Match**: `Query_ID 2`
- **Scores**:
  - Fuzzy Matching: Decent scores (76–83), showing it found a close match despite the typo.
  - TF-IDF: 0.707, showing a good match in vector space.
  - SBERT: 0.630, indicating decent semantic similarity.

**Conclusion**: All methods perform reasonably well, but Fuzzy Matching and SBERT slightly outperform TF-IDF.

#### **Row 4:**
- **Variation Query**: "Payment did not go through during chckout" (misspelled "checkout")
- **Best Match**: `Query_ID 2`
- **Scores**:
  - Fuzzy Matching: Scores in the 60–70 range, reflecting the impact of the typo.
  - TF-IDF: 0.707, good match in vector space.
  - SBERT: 0.452, lower score due to the typo affecting semantic matching.

**Conclusion**: Fuzzy Matching handles the typo better than SBERT, while TF-IDF performs similarly to Row 3.

#### **Row 5:**
- **Variation Query**: "Payment issue at check out"
- **Best Match**: `Query_ID 2`
- **Scores**:
  - Fuzzy Matching: Lower scores (57–71) than previous payment queries.
  - TF-IDF: 0.500, moderate performance.
  - SBERT: 0.830, very strong semantic match.

**Conclusion**: SBERT outperforms here due to its strong semantic understanding, while Fuzzy Matching and TF-IDF are moderate.

#### **Row 6:**
- **Variation Query**: "Application crashes when opening setings" (misspelled "settings")
- **Best Match**: `Query_ID 3`
- **Scores**:
  - Fuzzy Matching: Scores around 88, good match despite the typo.
  - TF-IDF: 0.775, strong match.
  - SBERT: 0.772, strong semantic understanding.

**Conclusion**: All methods perform well, with SBERT showing good semantic understanding despite the typo.

#### **Row 7:**
- **Variation Query**: "App crash when going to settings"
- **Best Match**: `Query_ID 3`
- **Scores**:
  - Fuzzy Matching: Scores in the high 80s, indicating strong performance.
  - TF-IDF: 0.722, good match.
  - SBERT: 0.914, the highest score indicating excellent semantic matching.

**Conclusion**: SBERT performs best, showcasing its ability to understand the intent of the query.

#### **Row 8:**
- **Variation Query**: "Settings cause the app to chrash" (misspelled "crash")
- **Best Match**: `Query_ID 3`
- **Scores**:
  - Fuzzy Matching: Scores are moderate (~40–65) due to the typo.
  - TF-IDF: 0.508, indicating a weaker match.
  - SBERT: 0.583, moderate semantic understanding.

**Conclusion**: All methods struggle with the typo, but SBERT still gives a reasonably good score compared to TF-IDF and Fuzzy Matching.

#### **Row 9:**
- **Variation Query**: "Forgot passwrd and cant reset" (misspelled "password")
- **Best Match**: `Query_ID 4`
- **Scores**:
  - Fuzzy Matching: Scores in the 70s, showing decent performance despite the typo.
  - TF-IDF: 0.783, strong match.
  - SBERT: 0.575, lower due to the typo.

**Conclusion**: Fuzzy Matching and TF-IDF handle the typo better, but SBERT's semantic understanding is affected by the misspelling.

---

### **General Insights**:
- **Fuzzy Matching** is effective at handling typos and minor misspellings but may struggle with more complex semantic differences.
- **TF-IDF** performs well on well-formed queries but struggles with misspellings or varying word choices.
- **SBERT** shines when queries are semantically similar, but its performance can drop significantly with spelling errors or typos.

For optimal matching performance, combining the strengths of all three methods could yield the best results, especially in handling both typographical errors and semantic variations.